server/license: Change how trial license usage is tracked

spilchen opened the pull request:
There’s a race condition when updating the enterprise.license config setting and checking the trial usage count. If a node starts up while a new trial license is being applied, it can encounter an issue where it sees the updated trial usage count in KV before receiving the corresponding enterprise.license config setting. This causes the license update to be rejected, as it incorrectly assumes a trial license has already been used.

This change addresses the issue by modifying what is stored in the KV for the trial license. Instead of tracking the number of trial licenses used, which would ever be 0 or 1, we now store the expiry timestamp of any active or past trial license. The enterprise.license validation function will compare the expiry of the new license against the cached value from KV. If the expiry timestamp is not set or matches the expiry of the new license, the validation will proceed. Otherwise, it will fail as before.

This change will be backported to 24.2, 24.1, 23.2 and 23.1.

Epic: CRDB-39988
Closes #131968
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132177)
<!-- Reviewable:end -->

rafiss reviewed (approved):
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132177)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/server/license/enforcer.go` line 552 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132177#-O8hDzal1JvgewH6PZ-d:-O8hDzal1JvgewH6PZ-e:b-74ccrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e5e9d0b2c89970ee816218a1ef786fc230655/pkg/server/license/enforcer.go#L552)):*
> ```Go
> // ResetTrialUsage is an API to clear the license enpiry in the KV. This is only used
> // for test purposes.
> func (e *Enforcer) ResetTrialUsage(ctx context.Context) error {
> ```

nit: we often name functions that are only for testing, but need to be exported with a prefix of `Testing`. i.e. `TestingResetTrialUsage`


<!-- Sent from Reviewable.io -->

spilchen committed:
server/license: Change how trial license usage is tracked

There’s a race condition when updating the enterprise.license config setting
and checking the trial usage count. If a node starts up while a new trial
license is being applied, it can encounter an issue where it sees the updated
trial usage count in KV before receiving the corresponding enterprise.license
config setting. This causes the license update to be rejected, as it
incorrectly assumes a trial license is already in use.

This change addresses the issue by modifying what is stored in the KV for the
trial license. Instead of tracking the number of installed trial licenses, we
now store the expiry timestamp of any active or past trial license. The
enterprise.license validation function will compare the expiry of the new
license against the cached value from KV. If the expiry timestamp is not set or
matches the expiry of the new license, the validation will proceed. Otherwise,
it will fail as before.

Epic: CRDB-39988
Closes #131968
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132177)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/server/license/enforcer.go` line 552 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132177#-O8hDzal1JvgewH6PZ-d:-O8hJsuN0ZJyMIVJOk3n:bykbvrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7e5e9d0b2c89970ee816218a1ef786fc230655/pkg/server/license/enforcer.go#L552)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we often name functions that are only for testing, but need to be exported with a prefix of `Testing`. i.e. `TestingResetTrialUsage`
</blockquote></details>

Good idea. That makes it clear.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR @rafiss 

bors r=rafiss
